File: atcoder/abc192/abc192_d.cpp
//
//  abc192_d.cpp
//  Problems
//
//  Created by NicolasCardozo on 4/6/21.
//  Copyright © 2021 NicolasCardozo. All rights reserved.
//

#include <stdio.h>
#include <bits/stdc++.h>

using namespace std;

string x;
long long m;
vector<int> arr;

long long f(long long base){
    long long res = 0;
    for(char c : x){
        if(res > m / base) return 0;
        res *= base;
        res += c;
    }
    return res <= m;
}

int main() {
    cin >> x;
    cin >> m;
    
    for(char& c : x) c -= '0';
    if (x.size() == 1)
        cout << f(10) << endl;
    else {
        long long d = *max_element(x.begin(), x.end());
        long long low = d;
        long long high = 1.1e18;
        while (high - low > 1) {
            long long mid = (low + high) / 2;
            if (f(mid))
                low = mid;
            else
                high = mid;
        }
        cout << low - d << endl;
    }
    return 0;
}

File: atcoder/abc192/abc192_d.kt
package atcoder

import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.*

fun main() {
    val In = BufferedReader(InputStreamReader(System.`in`))
    val x = In.readLine()!!
    val m = In.readLine()!!.toDouble()
    val arr = x.toCharArray()
    fun f(base: Double): Double {
        var res = 0.0
        val length = x.length
        for (i in 0 until length) {
            val value = arr[i] - '0'
            res += value * base.pow(length - 1 - i.toDouble())
        }
        return res
    }
    val d = x.toList().map { it.toString().toInt() }.reduce { acc, i -> max(i, acc) }
    if (x.length == 1 || m < d.toDouble())
        println(0)
    else {
        var low = d + 1.0
        var high = m + 1 
        var mid: Double
        while (high - low > 1) {
            mid = low + (high - low) / 2
            if (f(mid) <= m)
                low = mid
            else
                high = mid
        }
        println(low - d)
    }
}
